### PR TITLE
add hints for mips registers that may need restore

### LINKED ISSUE #1722: MIPS RSA pk encryption produces invalid result
May be related to #1093 and seen impacting big and little endian MIPS

### Description
- Type: Bug 
- Priority: Major | Minor
Mitigated by following suggestion to remove embedded assembly in `bn_mul.h` based on https://github.com/ARMmbed/mbedtls/issues/755#issuecomment-275922927

---------------------------------------------------------------
## Bug

**OS**  
Mbed OS linux MIPS and MIPSEL

**mbed TLS build:**  
Version: 2.6.0 and still present in 2.9.0
OS version: reproduced in `qemu-mips` and Openwrt 15.05 (Broadcom BCM5300)
Configuration: attached config.h
[config.h.txt](https://github.com/ARMmbed/mbedtls/files/2095231/config.h.txt)

Compiler and options (if you used a pre-built binary, please indicate how you obtained it):  
https://bitbucket.org/GregorR/musl-cross

**Peer device TLS stack and version**  
N/A failure can be reproduced in sample code, although this was identified by encrypting data with an RSA public key on MIPS and attempting to decrypt in Ruby 2.5.1 on x86_64.
RSA key was generated with `OpenSSL 1.0.2j`
```
openssl req -new -newkey rsa:2048 -nodes -x509 -subj '/CN=localhost' -days 365 -keyout server.key -out server.crt && cat server.key server.crt > server.pem
openssl rsa -in server.pem -pubout > server.pub
```

**Expected behavior**   
Sample code should report success on decrypt of public key encrypted data.
```
qemu-mips ./test
loaded public key
ALL IS WELL IN LOCAL DECRYPT
```

**Actual behavior**  
Decryption fails with error code
```
qemu-mips ./test
loaded public key
FAILED TO DECRYPT mbedtls_pk_decrypt RETURNED -16640
```

**Steps to reproduce**  
Sample failing code on MIPS, this code reports success on ARM & x86_64
```
#include <stdlib.h>
#include <mbedtls/aes.h>
#include <mbedtls/entropy.h>
#include <mbedtls/ctr_drbg.h>
#include <mbedtls/pk.h>

int main (argv)
{
	size_t olen = 0;
	size_t data_len = 16;
	unsigned char data[16] = "ABCDEFGHIJKLMNO";
	mbedtls_pk_context pk;
	mbedtls_ctr_drbg_context ctr_drbg;
	mbedtls_entropy_context entropy;

	mbedtls_entropy_init(&entropy);
	mbedtls_ctr_drbg_init(&ctr_drbg);

	mbedtls_pk_init(&pk);
	if (!(mbedtls_pk_parse_public_keyfile( &pk, "server.pub" ))) {
		printf("loaded public key\n");
		if (!(mbedtls_ctr_drbg_seed(&ctr_drbg, mbedtls_entropy_func, &entropy, NULL, 0))) {
			mbedtls_pk_context priv_k;
			mbedtls_pk_init(&priv_k);
			unsigned char buf[MBEDTLS_MPI_MAX_SIZE] = {'\0'};
			if (!(mbedtls_pk_encrypt(&pk, data, data_len, buf, &olen, sizeof(buf), mbedtls_ctr_drbg_random, &ctr_drbg))) {
				// now try to decrypt
				if (!(mbedtls_pk_parse_keyfile( &priv_k, "server.key", "" ))) {
					size_t dec_len = 0;
					unsigned char result[MBEDTLS_MPI_MAX_SIZE] = {'\0'};
					int ret = 0;
					if(!(ret = mbedtls_pk_decrypt( &priv_k, buf, olen, result, &dec_len, sizeof(result),
										mbedtls_ctr_drbg_random, &ctr_drbg)))
					{
						printf("ALL IS WELL IN LOCAL DECRYPT\n");
					}
					else
					{
						printf("FAILED TO DECRYPT mbedtls_pk_decrypt RETURNED %d\n", ret);
					}
				}
				else
				{
					printf("FAILED TO PARSE THE PRIVATE KEY\n");
				}
 			} else {
				printf("FAILED TO ENCRYPT DATA\n");
			}
		}

	} else
	{
		printf("FAILED TO PARSE THE PUBLIC KEY\n");
	}
	mbedtls_ctr_drbg_free(&ctr_drbg);
	mbedtls_entropy_free(&entropy);
        return 0;
}
```

compiled with sample `mips-linux-gnu-gcc -static main.c -o test -I"../mbedtls/include/" -L"../lib" -lmbedtls -lmbedcrypto` using https://hub.docker.com/r/asmimproved/qemu-mips/

### PR DESCRIPTION
**Pull request by jmartin-r7**
Fixes #1722, All credit given to @aurel32 as [original provider of this fix](https://github.com/ARMmbed/mbedtls/issues/1722#issuecomment-397112083), I simply consumed and validated in a project use case, on hardware and in the qemu emulation environment. 

## Description
Provide additional register notification to the compile for usage with the embedded MIPS assembly.

## Status
**READY**

## Requires Backporting
When there is a bug fix, it should be backported to all maintained and supported branches.
Yes

Which branch?  All currently supported branches *should* be safe to cherry-pick this change.

## Additional comments
Any additional information that could be of interest

## Todos
- [x] Tests (existing test *should* exercise this however there is no harness to currently execute all tests on MIPS environments.)
- [x] Changelog updated
- [x] Backported

## Steps to test or reproduce
Testing and reproduction for the original issue can be found in issue #1722

### CODE REVIEW

File: include/mbedtls/bn_mul.h [hunk 1]
```diff
@@ -734,7 +734,7 @@
         "sw     $10, %2         \n\t"   \
         : "=m" (c), "=m" (d), "=m" (s)                      \
         : "m" (s), "m" (d), "m" (c), "m" (b)                \
-        : "$9", "$10", "$11", "$12", "$13", "$14", "$15"    \
+        : "$9", "$10", "$11", "$12", "$13", "$14", "$15", "$lo", "$hi" \
```
Comment (segevfiner) [inline include/mbedtls/bn_mul.h]: Older gcc versions want `lo`/`hi` without a dollar. I don't remember which versions exactly.

**Comment by RonEld** [comment]
@jmartin-r7 Thank you for the contribution!
Although, as you mentioned, @aurel32 is the provider of this fix, and should be credited, you are still the author of this fix. As such, we would kindly ask you to accept our CLA.

Unfortunately our policy is to not accept contributions, without a Contributor’s Licence Agreement (CLA) signed or authorised by yourself or your employer.

If this is a personal contribution, the easiest way to do this is if you create an mbed account and [accept this click through agreement](https://os.mbed.com/contributor_agreement/). Alternatively, you can find a [slightly different agreement to sign here](https://www.mbed.com/en/about-mbed/contributor-license-agreements/), which can be signed and returned to us, and is applicable if you don't want to create an mbed account or alternatively if this is a corporate contribution.

Thanks for your understanding and again, thanks for the contribution!

**Comment by jmartin-r7** [comment]
I accepted this although the username is different I used the same contributor email address.

> You accepted the Contributor Agreement on Tue 14 Aug 2018.

**Comment by RonEld** [comment]
@jmartin-r7 Thank you for accepting the CLA!
Unfortunately, we can't link github and Mbed accounts, and we need to manually search your username in Mbed.
Could you please refer us to your Mbed account for confirmation?

**Comment by jmartin-r7** [comment]
Mbed account username is `Op3n4M3` email address matches to my git commit.

**Comment by jmartin-r7** [comment]
I'll look into request this week.

I will tests changing references on `hi, lo`, however I have yet to locate documentation on the gcc expected format.

Backports should be fairly straight forward will open PRs and reference this when I can complete testing.

**Comment by segevfiner** [comment]
I didn't find any documentation either. I had to look at gcc's code and use trial and error. :persevere:

**Comment by hanno-arm** [comment]
@jmartin-r7 Sounds good! :+1: If you are short on time, please let us know, and we'll happily finish off the PR for you and provide the backports. It would be nice to get this fix in soon.

**Comment by hanno-arm** [comment]
@jmartin-r7 Thanks for making the changes. I looked at the Travis Failure Log, and the failure is in a test where a proxy is randomly simulating package loss and reordering. It turned out that the proxy picked an unlucky seed leading almost all packages being dropped - the observed timeout is expected in this case (though one might argue that the CI should be able to not show this as a failure then).

File: ChangeLog [hunk 2]
```diff
@@ -11,6 +11,10 @@ Bugfix
      previously lead to a stack overflow on constrained targets.
    * Add `MBEDTLS_SELF_TEST` for the mbedtls_self_test functions
      in the header files, which missed the precompilation check. #971
+   * Extend register cleanup needed on MIPS assembly optimizations. Adds hi, lo
```
[the remark this answers is not in the document]
Comment (jmartin-r7) [reply ChangeLog]: Sounds good, updated.

**Comment by hanno-arm** [comment]
@aurel32 Please let us know how you'd like to be credited in the ChangeLog, as the PR's ready to go apart from that. Thanks!

**Comment by aurel32** [comment]
> @aurel32 Please let us know how you'd like to be credited in the ChangeLog, as the PR's ready to go apart from that. Thanks!

I don't have strong preference. Looking at the existing ChangeLog, it looks like people are mostly credited through their full name, so I only have a small preference for that, but I am fine with both options.

**Comment by aurel32** [comment]
> @aurel32 Could you provide us with your full name then?

Aurelien Jarno

**Comment by hanno-arm** [comment]
@aurel32 Thanks!

@jmartin-r7 Could you adapt the ChangeLog here and in the backports accordingly?

**Comment by sbutcher-arm** [comment]
Reviews approved and passes CI, so 'ready for merge'.

**Pull request merged**